service/node.go

```diff
@@ -504,9 +519,110 @@ func (s *service) NodeGetInfo(
 	}, nil
 }
 
-func (s *service) NodeGetVolumeStats(
-	ctx context.Context, req *csi.NodeGetVolumeStatsRequest) (*csi.NodeGetVolumeStatsResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+// NodeGetVolumeStats will check the status of a volume given its ID and path
+// if volume is healthy, stats on volume usage will be returned
+// if volume is unhealthy, a message will be returned detailing the issue
+// To determine if volume is healthy, this method checks: volume known to array, volume known to SDC, volume path readable, and volume path mounted
+// Note: kubelet only calls this method when feature gate: CSIVolumeHealth=true
+func (s *service) NodeGetVolumeStats(ctx context.Context, req *csi.NodeGetVolumeStatsRequest) (*csi.NodeGetVolumeStatsResponse, error) {
+
+	csiVolID := req.GetVolumeId()
+	volPath := req.GetVolumePath()
+	mounted := false
+	healthy := true
+	message := ""
+
+	//validate params first, make sure neither field is empty
+	if len(volPath) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "no volume Path provided")
+	}
+
+	if len(csiVolID) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "no volume ID  provided")
+	}
+
+	//check if volume exists
+
+	volID := getVolumeIDFromCsiVolumeID(csiVolID)
+	systemID := s.getSystemIDFromCsiVolumeID(csiVolID)
+
+	if systemID == "" {
+		// use default system
+		systemID = s.opts.defaultSystemID
+	}
+
+	_, err := s.getSDCMappedVol(volID, systemID, 30)
+	if err != nil {
+		//volume not known to SDC, next check if it exists at all
+		vol, _, err := s.listVolumes(systemID, 0, 0, true, false, volID, "")
+
+		if err != nil || len(vol) == 0 || vol[0].ID != volID {
+			return nil, status.Errorf(codes.NotFound, "volume with ID '%s' not found on array %s", volID, systemID)
```

Comment:
volume is not mapped not in array , however its known to k8s , better to return not healthy , instead of nil response.

Reply:
If the volume is not on array, nor is it on the node, it does not exist. Spec says to return error when volume does not exist.

Reply:
if a volume is in k8s  but not on array its abnormal , return err is in appropriate

Reply:
We don't know if volume is in k8s at this stage. All we know is that kubelet called NodeGetStats with the volume. 
Since we're dealing with array based storage, if array does not have volume and node does not have volume, kubelet has made a bad call and we need to return error

Reply:
"kubelet has made a bad call " --we should not code with this assumption

for both volume not mapped and volume not found - we must set healthy=false and return response object. refer csi slack discussion both controller and node get vol must set abnormal true.
This will make the health monitor put an event in POD . customer needs this event,

the code in the PR returns error, for vol not found, this will not get recorded in the event . hence we should not return error.

also if volume is not mapped and volume is found  , must return , skip rest of the code with abnormal =true.

Reply:
changed to return abnormal if vol not found on array